### PR TITLE
Code refactoring in subject area. And fixed bug for find all categories without request params.

### PR DESCRIPTION
**Pull request by immmus**
Code refactoring in subject area. I brought the code to a more readable form. Also, along the way, I fixed some shortcomings in the form of calling deprecated methods and the like.

Removed unnecessary double "new HashSet". But I’m still worried that these are mutable collections and that in 95% of cases they’re simply not usedRemoved unnecessary double "new HashSet". But I’m still worried that these are mutable collections and that in 95% of cases they’re simply not used

### CODE REVIEW

**Comment by immmus** [comment]
In case of commit  https://github.com/odpi/egeria/pull/3050/commits/d0860ebe3897c4860c3413736290cab5eb00b044

I noticed that the toString method returns a StringBuilder which is not correct based on the logic. I decided to correct this defect, but I realized that a lot of architectural logic was already tied to this method. You can fix it in absolutely all classes and simultaneously remove a lot of duplicated code, but I was worried about the question "Will there be a backward compatibility problem?". So I rolled back all my changes, but missed it.

**Comment by immmus** [comment]
Before fixing the bug, when requesting all categories without parameters to

_http://localhost:8080/servers/test/open-metadata/access-services/subject-area/users/admin/categories_
```

I was get next exception:
{
    "class": "InvalidParameterExceptionResponse",
    "relatedHTTPCode": 200,
    "responseCategory": "InvalidParameterException",
    "exceptionClassName": "org.odpi.openmetadata.accessservices.subjectarea.server.handlers.ErrorHandler",
    "exceptionErrorMessage": "OMAS-SUBJECT-AREA-400-035 Invalid parameter processing request findCategory.",
    "exceptionSystemAction": "The system is unable to process the request.",
    "exceptionUserAction": "Please correct the parameter and retry."
}
```

After I fixed the bug and i've filling catalog via org.odpi.openmetadata.accessservices.subjectarea.samples.SubjectAreaDefinitionSample.
And I was get next the picture when I was called all categories again, I get only one entity instead of all and with each request different.

The fact is that the findCategory method in the class org.odpi.openmetadata.accessservices.subjectarea.handlers.SubjectAreaCategoryHandler 
has the following limitation:

```
response = getCategory(userId, entityDetail.getGUID());
                    if (response.getResponseCategory() == ResponseCategory.Category) {
                        CategoryResponse categoryResponse = (CategoryResponse) response;
                        Category category = categoryResponse.getCategory();
                        categories.add(category);
                    } else {
````

That is, if an entity arrives with a different ResponseCategory other than Category, the method does not add (in this case, SubjectAreaDefinition).
The problem is clearly of an architectural nature, as when adding any new ResponseCategory, the method will grow "if else", which will make the code less supported. The architecture of work in this case is worth rethinking.

This problem is absent if you simply use the Category entity, so I changed the logic of SubjectAreaDefinitionSample in my next commit to have a working example.

Thx. I hope for a speedy review, because current error is present in production.

**Comment by immmus** [comment]
I also noticed that in the class org.odpi.openmetadata.accessservices.subjectarea.server.services.SubjectAreaRESTServicesInstance there are two big unused methods "getRelationshipsFromGuid" written a year ago

**Comment by davidradl** [comment]
@immmus Thanks for your pr. I am finding it difficult to review this change as there is a lot of reformatting. It would really help if you could split this up into small minimal changes each with an issue and pr. If there are refactors then an issue and pr for just that refactor. A targeted fix on the bug in master you talk about could be quickly reviewed if that fix were by itself.

**Comment by immmus** [comment]
> @immmus Thanks for your pr. I am finding it difficult to review this change as there is a lot of reformatting. It would really help if you could split this up into small minimal changes each with an issue and pr. If there are refactors then an issue and pr for just that refactor. A targeted fix on the bug in master you talk about could be quickly reviewed if that fix were by itself.

Hi, if you say about request for search all categories then it here https://github.com/odpi/egeria/pull/3050/commits/0d0d7752568adc4540a4e6f64097f9ed4b7a4bf4

**Comment by immmus** [comment]
> I am not sure this fix is correct, the change means that sequencingOrder and sequencingProperty will not be applied when there is no searchCriteria. Also all API changes need to be tested in the SubjectAreaFVT. SubjectAreaFVT tests all the APIs that are known to work.
> 
> I am thinking
> 
> * I suspect if we specify .* when we are requested with null as the searchCriteria, we should then be able to use the findEntitiesByPropertyValue and the sequencingOrder and Property.
> * I have not looked into what searchCriteria of null means to OMRS, whether this means it looks for null values or unspecified values. The Conformance tests may/should check the behaviour; I have not investigated.
> * I suspect an equivalent change would be need to Project Glossary and Term processing



Yes, to search for all Terms, Glossaries and Projects, the same logic of work is used. And in my correction, I relied on their logic of work, considering it correct. The searchCriteria of null means that we are looking for everything without any search criteria, as I understand it.

If null searchCriteria, sequencingOrder and sequencingProperty should be used. Then the logic of work for all of them is written and works incorrectly.

By the way, I ran the tests, they all passed. After all, they absolutely do not check the logic of work for sequencingOrder and sequencingProperty as it turned out.

File: open-metadata-resources/open-metadata-samples/access-services-samples/subject-area-client-samples/pom.xml [hunk 1]
```diff
@@ -62,7 +62,7 @@
                             <archive>
                                 <manifest>
                                     <mainClass>
-                                        org.odpi.openmetadata.accessservices.subjectarea.samples.SubjectAreaDefinitionSample
+                                        org.odpi.openmetadata.accessservices.subjectarea.samples.SubjectAreaCatalogSample
```
Comment (davidradl) [inline open-metadata-resources/open-metadata-samples/access-services-samples/subject-area-client-samples/pom.xml:65]: I am unsure why this is being changed

File: open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/handlers/SubjectAreaCategoryHandler.java [hunk 2]
```diff
@@ -317,7 +316,12 @@ public SubjectAreaOMASAPIResponse findCategory(String userId,
         SubjectAreaCategoryRESTServices categoryRESTServices = new SubjectAreaCategoryRESTServices();
         categoryRESTServices.setOMRSAPIHelper(this.oMRSAPIHelper);
 
-        response = this.oMRSAPIHelper.findEntitiesByPropertyValue(methodName, userId, "GlossaryCategory", searchCriteria, asOfTime, offset, pageSize, sequencingOrder, sequencingProperty, methodName);
+        if (searchCriteria == null) {
+            response = oMRSAPIHelper.getEntitiesByType(oMRSAPIHelper, methodName, userId, "GlossaryCategory", asOfTime, offset, pageSize);
```
Comment (davidradl) [inline open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/handlers/SubjectAreaCategoryHandler.java:320]: This should not use getEntitiesByType, it needs to use findEntitiesByPropertyValue so the sequencing parameters can be honoured.

File: open-metadata-resources/open-metadata-samples/access-services-samples/subject-area-client-samples/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/samples/SubjectAreaCatalogSample.java [hunk 3]
```diff
@@ -17,9 +17,9 @@
 import java.io.InputStreamReader;
 
 /**
- *  SubjectAreaDefinition Sample provides a client program that calls the Subject Area OMAS to create sample Subject Area Definitions.
+ *  Category Sample provides a client program that calls the Subject Area OMAS to create sample Subject Area Catalog.
  */
-public class SubjectAreaDefinitionSample
```
Comment (davidradl) [inline open-metadata-resources/open-metadata-samples/access-services-samples/subject-area-client-samples/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/samples/SubjectAreaCatalogSample.java:22]: This class should not be renamed. it is not a subject Area catalog sample.

File: open-metadata-resources/open-metadata-samples/access-services-samples/subject-area-client-samples/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/samples/SubjectAreaCatalogSample.java [hunk 4]
```diff
@@ -47,7 +47,7 @@
     private static final String   DISTRIBUTION = "Distribution";
     private static final String   INVOICE = "Invoice";
 
-    private static final String GLOSSARY_NAME        = "Coco Pharmaceuticals Subject Area Definitions";
```
Comment (davidradl) [inline open-metadata-resources/open-metadata-samples/access-services-samples/subject-area-client-samples/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/samples/SubjectAreaCatalogSample.java:50]: This change is incorrect

File: open-metadata-resources/open-metadata-samples/access-services-samples/subject-area-client-samples/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/samples/SubjectAreaCatalogSample.java [hunk 5]
```diff
@@ -67,7 +67,7 @@
      * @param serverName server to connect to.
      * @param clientUserId user id to use to access metadata.
      */
-    private SubjectAreaDefinitionSample(String serverURLRoot, String serverName, String clientUserId)
```
Comment (davidradl) [inline open-metadata-resources/open-metadata-samples/access-services-samples/subject-area-client-samples/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/samples/SubjectAreaCatalogSample.java:70]: should not be renamed

File: open-metadata-resources/open-metadata-samples/access-services-samples/subject-area-client-samples/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/samples/SubjectAreaCatalogSample.java [hunk 6]
```diff
@@ -93,92 +93,92 @@ private void run() throws SubjectAreaCheckedExceptionBase
          * create Organisation subject areas
          */
         String glossaryGuid =  glossary.getSystemAttributes().getGUID();
-        SubjectAreaDefinition organisation = createTopSubjectAreaDefinition(ORGANIZATION,glossaryGuid);
-        SubjectAreaDefinition hospital= createChildSubjectAreaDefinition(HOSPITAL,organisation,glossaryGuid);
-        SubjectAreaDefinition supplier= createChildSubjectAreaDefinition(SUPPLIER,organisation,glossaryGuid);
+        Category organisation = createTopCategory(ORGANIZATION,glossaryGuid);
```
Comment (davidradl) [inline open-metadata-resources/open-metadata-samples/access-services-samples/subject-area-client-samples/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/samples/SubjectAreaCatalogSample.java:96]: This sample is creating subject area definitions not categories - these renames of the types are incorrect

File: open-metadata-resources/open-metadata-samples/access-services-samples/subject-area-client-samples/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/samples/SubjectAreaCatalogSample.java [hunk 7]
```diff
@@ -93,92 +93,92 @@ private void run() throws SubjectAreaCheckedExceptionBase
...
-     * @param name name of the Subject Area Definition to create
+     * @param name name of the Subject Area Category to create
      * @param parent parent Category
      * @param glossaryGuid guid of the glossary to associate the Subject Area with.
-     * @return SubjectAreaDefinition the created Subject Area Definition
+     * @return Category the created Subject Area
      * @throws SubjectAreaCheckedExceptionBase error
      */
-    private SubjectAreaDefinition createChildSubjectAreaDefinition(String name, SubjectAreaDefinition parent, String glossaryGuid) throws SubjectAreaCheckedExceptionBase
+    private Category createChildCategory(String name, Category parent, String glossaryGuid) throws SubjectAreaCheckedExceptionBase
```
Comment (davidradl) [inline open-metadata-resources/open-metadata-samples/access-services-samples/subject-area-client-samples/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/samples/SubjectAreaCatalogSample.java:165]: all these renames are incorrect

**Comment by davidradl** [comment]
Please can you split this into separate issues and have pr for each:
-one for the reformats (not refactor) 
- one for the search issue
- one for the warnings around hashmaps.
- any others?
It would be so much easier to review and handle.

**Comment by immmus** [comment]
> Please can you split this into separate issues and have pr for each:
> -one for the reformats (not refactor)
> 
> * one for the search issue
> * one for the warnings around hashmaps.
> * any others?
>   It would be so much easier to review and handle.

Well, I’ll try to scatter all these tasks on different pr, it will take time and this pr will need to be closed

**Comment by davidradl** [comment]
@immmus  I am actually in the middle of refactoring subject area around the error messages to change the SubjectAreaError processing to use the new MessageDefinitions as part of #3070. I am going to complete this refactor and push that change (I will try to remember to format the file after my changes :-) - as I have left in a few formatting errors - which you have rightly picked up on); I will then endeavour to work with you so we can make the appropriate changes for the issues you have raised.

**Comment by immmus** [comment]
> @immmus I am actually in the middle of refactoring subject area around the error messages to change the SubjectAreaError processing to use the new MessageDefinitions as part of #3070. I am going to complete this refactor and push that change (I will try to remember to format the file after my changes :-) - as I have left in a few formatting errors - which you have rightly picked up on); I will then endeavour to work with you so we can make the appropriate changes for the issues you have raised.

I made all my changes in separate pr as you asked. Waiting for your approve.
https://github.com/odpi/egeria/pull/3071
https://github.com/odpi/egeria/pull/3069
https://github.com/odpi/egeria/pull/3067

**Comment by immmus** [comment]
@davidradl  I can to close this? You can are checking my new pr's

**Pull request closed without merging**